lib/autoupdate/tools/updater.go

```diff
@@ -0,0 +1,391 @@
+	if err := packaging.ReplaceToolsBinaries(u.toolsDir, archivePath, extractDir, u.tools); err != nil {
+		return trace.Wrap(err)
+	}
+	// Cleanup the tools directory with previously downloaded and un-archived versions.
+	if err := packaging.RemoveWithSuffix(u.toolsDir, updatePackageSuffix, pkgName); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// Exec re-executes tool command with same arguments and environ variables.
+func (u *Updater) Exec() (int, error) {
+	path, err := toolName(u.toolsDir)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+	// To prevent re-execution loop we have to disable update logic for re-execution.
+	env := append(os.Environ(), teleportToolsVersionEnv+"=off")
+
+	if runtime.GOOS == constants.WindowsOS {
+		cmd := exec.Command(path, os.Args[1:]...)
+		cmd.Env = env
+		cmd.Stdin = os.Stdin
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		if err := cmd.Run(); err != nil {
+			return 0, trace.Wrap(err)
+		}
+
+		return cmd.ProcessState.ExitCode(), nil
+	}
+
+	if err := syscall.Exec(path, append([]string{path}, os.Args[1:]...), env); err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	return 0, nil
+}
+
+func (u *Updater) downloadHash(ctx context.Context, url string) (string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	resp, err := u.client.Do(req)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == http.StatusNotFound {
+		return "", trace.NotFound("hash file is not found: %v", resp.StatusCode)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return "", trace.BadParameter("bad status when downloading archive hash: %v", resp.StatusCode)
+	}
+
+	var buf bytes.Buffer
+	_, err = io.CopyN(&buf, resp.Body, sha256.Size*2) // SHA bytes to hex
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	raw := buf.String()
+	if _, err = hex.DecodeString(raw); err != nil {
+		return "", trace.Wrap(err)
+	}
+	return raw, nil
+}
+
+func (u *Updater) downloadArchive(ctx context.Context, downloadDir string, url string) (string, string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+	resp, err := u.client.Do(req)
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == http.StatusNotFound {
+		return "", "", trace.NotFound("archive file is not found: %v", resp.StatusCode)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return "", "", trace.BadParameter("bad status when downloading archive: %v", resp.StatusCode)
+	}
+
+	if resp.ContentLength != -1 {
+		if err := checkFreeSpace(u.toolsDir, uint64(resp.ContentLength)); err != nil {
+			return "", "", trace.Wrap(err)
+		}
+	}
+
+	// Caller of this function will remove this file after the atomic swap has
+	// occurred.
+	f, err := os.CreateTemp(downloadDir, "tmp-")
```

Comment:
This file descriptor is leaked. It's impossible to close after the method returns, and writing may not actually be complete. This needs a close with an error check.

Also, I'd suggest passing the file path (or maybe just a Writer) to write to as an argument, so that the caller knows that it must delete the file if the method returns an error, and additionally, deleting the file on error.

Example: https://github.com/gravitational/teleport/pull/47565/files#diff-8961b5cac012637602dcc182a8b40b169a527f663072dc34b0ab2db12922e86dR118-R127